engine/resources/docker.go

```diff
@@ -0,0 +1,292 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+)
+
+func init() {
+	engine.RegisterResource("docker", func() engine.Res { return &DockerRes{} })
+}
+
+// DockerRes is a docker account resource.
+type DockerRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Edgeable
+
+	Image string            `yaml:"image"`
+	Cmd   []string          `yaml:"cmd"`
+	Env   []string          `yaml:"env"`
+	Ports map[string]string `yaml:"ports"`
+
+	init   *engine.Init
+	client *client.Client
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *DockerRes) Default() engine.Res {
+	return &DockerRes{}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *DockerRes) Validate() error {
+	return nil
+}
+
+// Init runs some startup code for this resource.
+func (obj *DockerRes) Init(init *engine.Init) error {
+	var err error
+	obj.init = init // save for later
+
+	obj.client, err = client.NewClient("unix:///var/run/docker.sock", "v1.37", nil, nil)
+	if err != nil {
+		return fmt.Errorf("error creating docker client: %s", err)
+	}
+
+	return nil
+}
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *DockerRes) Close() error {
+	return nil
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *DockerRes) Watch() error {
+	ctx := context.TODO()
+
+	eventChan, errChan := obj.client.Events(ctx, types.EventsOptions{})
+
+	// notify engine that we're running
+	if err := obj.init.Running(); err != nil {
+		return err // exit if requested
+	}
+
+	var send = false // send event?
+	for {
+		if obj.init.Debug {
+			obj.init.Logf("Watching: %s", passwdFile) // attempting to watch...
```

Comment:
Already changed. ;)

Reply:
you reviewed your own code, hehe, you can just push a new commit ;)